cep-abi3.md

```diff
@@ -0,0 +1,269 @@
+  They have `noarch: python` in `info/index.json`.
+
+&nbsp;&nbsp;<strong>A3</strong>:
+  Python files are in `<PREFIX>/site-packages`.
+
+&nbsp;&nbsp;<strong>A4</strong>:
+  Entry points are recorded in `info/link.json`.
+
+A conda install tool does four things to support them:
+
+&nbsp;&nbsp;<strong>B1</strong>:
+  Files in `<PREFIX>/site-packages` are moved to the correct location. Eg:
+  `<PREFIX>/lib/python3.10/site-packages`.
+
+&nbsp;&nbsp;<strong>B2</strong>:
+  python files (files ending with `*.py`) are compiled to `.pyc` files. Eg:
+  `<PREFIX>/lib/python3.10/site-packages/foo.py` is compiled to
+  `<PREFIX>/lib/python3.10/site-packages/__pycache__/foo.cpython-310.pyc`.
+
+&nbsp;&nbsp;<strong>B3</strong>:
+  `.pyc` files created are recorded in `<PREFIX>/conda-meta/<pkg>.json`
+  so that they are removed properly when the package is uninstalled.
+
+&nbsp;&nbsp;<strong>B4</strong>:
+  Entry points in `info/link.json` are materialised.
+
+### info/link.json file
+An example `info/link.json` for `noarch: python` looks like
+
+```json
+{
+  "noarch": {
+    "entry_points": [
+      "isympy = isympy:main"
+    ],
+    "type": "python"
+  },
+  "package_metadata_version": 1,
+  "preferred_env": "foo"
+}
+```
+
+An example for `info/link.json` for `noarch: generic` looks like
+```
+{
+  "noarch": {
+    "type": "generic"
+  },
+  "package_metadata_version": 1
+}
+```
+
+
+Here `preferred_env` is ignored by conda since 2017 and is not supported by
+other conda install tools. Therefore `info/link.json` is used exclusively
+for `noarch` packages and out of the two types, `noarch: generic` packages
+does not require any special action.
+
+### info/index.json file
+
+An example for a `noarch: python` recipe.
+
+```json
+{
+  "arch": null,
+  "build": "pyh2585a3b_103",
+  "build_number": 103,
+  "depends": [
+    "mpmath >=0.19",
+    "python >=3.8"
+  ],
+  "license": "BSD-3-Clause",
+  "license_family": "BSD",
+  "name": "sympy",
+  "noarch": "python",
+  "platform": null,
+  "subdir": "noarch",
+  "timestamp": 1718625708903,
+  "version": "1.12.1"
+}
+```
+
+### Current behaviour in solver tools
+
+Conda package upload tools like `anaconda-client` use `A1` to upload
+the package to the `noarch` subdir.
+
+Conda install tools have slightly different behaviour.
+
+Conda:
+1. Actions `B1, B2, B3` are applied for packages with `A3`.
+2. Action `B4` is applied for packages with `A4`.
+
+Micromamba:
+1. Actions `B1, B2, B3` are applied for packages with both `A2, A3`.
+2. Action `B4` is applied for packages with both `A2, A4`.
+
+
+## Implementation for abi3 packages in install tools.
+
+In order to support abi3 packages, we propose two methods.
+
+### `package_metadata_version = 1`
+
+We require the following attributes in abi3 packages:
+
+&nbsp;&nbsp;<strong>C1</strong>:
+  They have `subdir: <platform>` where `<platform>` is the subdir
+  that the package was built for.
+
+&nbsp;&nbsp;<strong>C2</strong>:
+  They have `noarch: python`.
```

Comment:
Should there be a suggestion for a new way to specify abi3 packages in this proposal? I understand how `noarch: python` almost works already and so can be used to get abi3 packages working with existing tools, but it is a misleading label since these packages are still architecture dependent.

Also, I had tried to test with `post-link.sh` and `pre-unlink.sh` scripts previously [here](https://github.com/conda-forge/qiskit-terra-feedstock/pull/41) but I didn't try `noarch: python` and got stuck on the run exports of Python. Maybe I just didn't know the right syntax, but it seemed like when you don't use `noarch: python` `conda-build` pins the Python version in a way that is hard to override by setting the Python version in the recipe requirements. That might be worth mentioning as another important behavior of `noarch: python` (if I was not missing something).

Reply:
> Should there be a suggestion for a new way to specify abi3 packages in this proposal? I understand how noarch: python almost works already and so can be used to get abi3 packages working with existing tools, but it is a misleading label since these packages are still architecture dependent.

This is `package_metadata_version = 1` which is for existing tool compatibility. Please see the next section for new ways.

> Also, I had tried to test with post-link.sh and pre-unlink.sh scripts previously https://github.com/conda-forge/qiskit-terra-feedstock/pull/41 but I didn't try noarch: python and got stuck on the run exports of Python. Maybe I just didn't know the right syntax, but it seemed like when you don't use noarch: python conda-build pins the Python version in a way that is hard to override by setting the Python version in the recipe requirements. That might be worth mentioning as another important behavior of noarch: python (if I was not missing something).

That should be fixable by using `ignore_run_exports`. This is only specific to conda-forge which has `run_exports` on `python` package and defaults does not AFAIK. I wanted to keep the CEP more generic.